engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -1,27 +1,35 @@
 package org.enso.interpreter.runtime.scope;
 
-import com.oracle.truffle.api.CompilerDirectives;
-
-import java.util.*;
-
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
 import com.oracle.truffle.api.interop.TruffleObject;
+
 import org.enso.interpreter.runtime.Context;
 import org.enso.interpreter.runtime.Module;
 import org.enso.interpreter.runtime.callable.function.Function;
 import org.enso.interpreter.runtime.data.Type;
 import org.enso.interpreter.runtime.error.RedefinedMethodException;
 import org.enso.interpreter.runtime.error.RedefinedConversionException;
 
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
 /** A representation of Enso's per-file top-level scope. */
 public final class ModuleScope implements TruffleObject {
   private final Type associatedType;
   private final Module module;
-  private Map<String, Object> polyglotSymbols = new HashMap<>();
-  private Map<String, Type> types = new HashMap<>();
-  private Map<Type, Map<String, Function>> methods = new HashMap<>();
-  private Map<Type, Map<Type, Function>> conversions = new HashMap<>();
-  private Set<ModuleScope> imports = new HashSet<>();
-  private Set<ModuleScope> exports = new HashSet<>();
+  private @CompilationFinal Map<String, Object> polyglotSymbols;
```

Comment:
I doubt `@CompilationFinal` will work for anything. First of all `ModuleScope` is unlikely to be known as constant (or am I wrong?). Second `java.util.HashMap` & co. are totally unsuitable for partial evaluation.

Reply:
OK. I will bring it back